src/recon_buildblock/LogcoshPrior.cxx

```diff
@@ -0,0 +1,666 @@
+      }
+    }
+  }
+
+  info(boost::format("Prior gradient max %1%, min %2%\n") % prior_gradient.find_max() % prior_gradient.find_min());
+
+  static int count = 0;
+  ++count;
+  if (gradient_filename_prefix.size()>0)
+  {
+    char *filename = new char[gradient_filename_prefix.size()+100];
+    sprintf(filename, "%s%d.v", gradient_filename_prefix.c_str(), count);
+    OutputFileFormat<DiscretisedDensity<3,elemT> >::default_sptr()->
+            write_to_file(filename, prior_gradient);
+    delete[] filename;
+  }
+}
+
+template <typename elemT>
+void
+LogcoshPrior<elemT>::
+compute_Hessian(DiscretisedDensity<3,elemT>& prior_Hessian_for_single_densel,
+                const BasicCoordinate<3,int>& coords,
+                const DiscretisedDensity<3,elemT> &current_image_estimate)
+{
+  assert(  prior_Hessian_for_single_densel.has_same_characteristics(current_image_estimate));
+  prior_Hessian_for_single_densel.fill(0);
+  if (this->penalisation_factor==0)
+  {
+    return;
+  }
+
+
+  const DiscretisedDensityOnCartesianGrid<3,elemT>& current_image_cast =
+          dynamic_cast< const DiscretisedDensityOnCartesianGrid<3,elemT> &>(current_image_estimate);
+
+  DiscretisedDensityOnCartesianGrid<3,elemT>& prior_Hessian_for_single_densel_cast =
+          dynamic_cast<DiscretisedDensityOnCartesianGrid<3,elemT> &>(prior_Hessian_for_single_densel);
+
+  if (weights.get_length() ==0)
+  {
+    compute_weights(weights, current_image_cast.get_grid_spacing(), this->only_2D);
+  }
+
+
+  const bool do_kappa = !is_null_ptr(kappa_ptr);
+
+  if (do_kappa && kappa_ptr->has_same_characteristics(current_image_estimate))
+    error("LogcoshPrior: kappa image has not the same index range as the reconstructed image\n");
+
+  const int z = coords[1];
+  const int y = coords[2];
+  const int x = coords[3];
+  const int min_dz = max(weights.get_min_index(), prior_Hessian_for_single_densel.get_min_index()-z);
+  const int max_dz = min(weights.get_max_index(), prior_Hessian_for_single_densel.get_max_index()-z);
+
+  const int min_dy = max(weights[0].get_min_index(), prior_Hessian_for_single_densel[z].get_min_index()-y);
+  const int max_dy = min(weights[0].get_max_index(), prior_Hessian_for_single_densel[z].get_max_index()-y);
+
+  const int min_dx = max(weights[0][0].get_min_index(), prior_Hessian_for_single_densel[z][y].get_min_index()-x);
+  const int max_dx = min(weights[0][0].get_max_index(), prior_Hessian_for_single_densel[z][y].get_max_index()-x);
+
+  elemT diagonal = 0;
+  for (int dz=min_dz;dz<=max_dz;++dz)
+    for (int dy=min_dy;dy<=max_dy;++dy)
+      for (int dx=min_dx;dx<=max_dx;++dx)
+      {
+        // (temp sech^2(|temp|/scalar)) / (|temp|)
+        // sech(temp) = 1/cosh(temp)
+        elemT temp = current_image_estimate[z][y][x] - current_image_estimate[z+dz][y+dy][x+dx];
+        elemT current = weights[dz][dy][dx] * (temp / fabs(temp)) *
+                square((1/ cosh(fabs(temp) / this->scalar)));
```

Comment:
this will run into trouble at `temp==0` as usual. another Taylor expansion?

Reply:
seems there's no C++ sign function weirdly enough. Could [write our own](https://stackoverflow.com/questions/1903954/is-there-a-standard-sign-function-signum-sgn-in-c-c), or use [std::copysign](http://www.cplusplus.com/reference/cmath/copysign/) although it might be cryptic, so would need doc
```
In any case, let's move this to its own function as well (I suggest Hessian as well, although maybe somewhat weird for a 1D function)
std::copysign(square((1/ cosh(fabs(temp) / this->scalar))), temp)

Reply:
The function was wrong due to the abs(). Instead, the Hessian is simply `sech^2(scalar * temp)`. This is a well behaved function